File: backend/app/services/dashboard_service.py
"""Dashboard aggregation service (Module 1)."""
from __future__ import annotations

from datetime import date, datetime, timedelta

from sqlalchemy import func, select

from app.extensions import db
from app.models import Car, Payment, Rental
from app.models.enums import CarStatus, RentalStatus


def _car_counts() -> dict[str, int]:
    stmt = select(Car.status, func.count(Car.id)).group_by(Car.status)
    counts = {status.value: 0 for status in CarStatus}
    for status, count in db.session.execute(stmt).all():
        key = status.value if isinstance(status, CarStatus) else status
        counts[key] = count
    return counts


def _revenue(start: datetime, end: datetime) -> float:
    stmt = select(func.coalesce(func.sum(Payment.amount), 0.0)).where(
        Payment.paid_at >= start, Payment.paid_at < end
    )
    return float(db.session.execute(stmt).scalar_one())


def get_summary() -> dict:
    """Return the headline dashboard metrics."""
    counts = _car_counts()

    today = date.today()
    day_start = datetime.combine(today, datetime.min.time())
    month_start = datetime.combine(today.replace(day=1), datetime.min.time())

    today_revenue = _revenue(day_start, day_start + timedelta(days=1))
    next_month = (month_start.replace(day=28) + timedelta(days=4)).replace(day=1)
    monthly_revenue = _revenue(month_start, next_month)

    active_rentals = db.session.execute(
        select(func.count(Rental.id)).where(Rental.status == RentalStatus.ACTIVE)
    ).scalar_one()

    upcoming_returns = db.session.execute(
        select(func.count(Rental.id)).where(
            Rental.status == RentalStatus.ACTIVE,
            Rental.rental_end <= today + timedelta(days=3),
        )
    ).scalar_one()

    return {
        "cars_available": counts.get(CarStatus.AVAILABLE.value, 0),
        "cars_rented": counts.get(CarStatus.RENTED.value, 0),
        "cars_reserved": counts.get(CarStatus.RESERVED.value, 0),
        "cars_maintenance": counts.get(CarStatus.MAINTENANCE.value, 0),
        "today_revenue": round(today_revenue, 2),
        "monthly_revenue": round(monthly_revenue, 2),
        "active_rentals": active_rentals,
        "upcoming_returns": upcoming_returns,
    }


def revenue_by_month(months: int = 12) -> list[dict]:
    """Return revenue totals grouped by calendar month."""
    today = date.today().replace(day=1)
    results: list[dict] = []
    cursor = today
    # Walk backwards `months` months.
    series: list[date] = []
    for _ in range(months):
        series.append(cursor)
        cursor = (cursor - timedelta(days=1)).replace(day=1)
    for month_start in reversed(series):
        next_month = (month_start.replace(day=28) + timedelta(days=4)).replace(day=1)
        amount = _revenue(
            datetime.combine(month_start, datetime.min.time()),
            datetime.combine(next_month, datetime.min.time()),
        )
        results.append({"month": month_start.strftime("%Y-%m"), "revenue": round(amount, 2)})
    return results


def top_rented_cars(limit: int = 5) -> list[dict]:
    """Return the most frequently rented cars."""
    stmt = (
        select(Car, func.count(Rental.id).label("rentals"))
        .join(Rental, Rental.car_id == Car.id)
        .group_by(Car.id)
        .order_by(func.count(Rental.id).desc())
        .limit(limit)
    )
    rows = db.session.execute(stmt).all()
    return [{"car": car.display_name, "rentals": count} for car, count in rows]


def rental_statistics() -> dict:
    """Return rental counts grouped by status."""
    stmt = select(Rental.status, func.count(Rental.id)).group_by(Rental.status)
    stats = {status.value: 0 for status in RentalStatus}
    for status, count in db.session.execute(stmt).all():
        key = status.value if isinstance(status, RentalStatus) else status
        stats[key] = count
    return stats

File: backend-go/internal/service/dashboard_service.go
package service

import (
	"fmt"
	"time"

	"github.com/instaagrammeta/rentacar/backend-go/internal/models"
	"github.com/instaagrammeta/rentacar/backend-go/internal/pricing"
)

// revenueBetween sums payment amounts in the half-open interval [start, end).
func (s *Service) revenueBetween(start, end time.Time) float64 {
	var total float64
	s.DB.Model(&models.Payment{}).
		Where("paid_at >= ? AND paid_at < ?", start, end).
		Select("COALESCE(SUM(amount), 0)").Scan(&total)
	return total
}

func startOfDay(t time.Time) time.Time {
	return time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, t.Location())
}

func startOfMonth(t time.Time) time.Time {
	return time.Date(t.Year(), t.Month(), 1, 0, 0, 0, 0, t.Location())
}

// DashboardSummary returns the headline dashboard metrics.
func (s *Service) DashboardSummary() map[string]interface{} {
	counts := map[string]int64{}
	for _, st := range models.AllCarStatuses() {
		counts[string(st)] = 0
	}
	type row struct {
		Status string
		Count  int64
	}
	var rows []row
	s.DB.Model(&models.Car{}).Select("status, count(id) as count").Group("status").Scan(&rows)
	for _, r := range rows {
		counts[r.Status] = r.Count
	}

	now := time.Now()
	dayStart := startOfDay(now)
	monthStart := startOfMonth(now)
	nextMonth := monthStart.AddDate(0, 1, 0)

	todayRevenue := s.revenueBetween(dayStart, dayStart.AddDate(0, 0, 1))
	monthlyRevenue := s.revenueBetween(monthStart, nextMonth)

	var activeRentals int64
	s.DB.Model(&models.Rental{}).Where("status = ?", models.RentalActive).Count(&activeRentals)

	var upcomingReturns int64
	limit := startOfDay(now).AddDate(0, 0, 4) // rental_end <= today+3 (dates)
	s.DB.Model(&models.Rental{}).
		Where("status = ? AND rental_end < ?", models.RentalActive, limit).
		Count(&upcomingReturns)

	return map[string]interface{}{
		"cars_available":   counts[string(models.CarAvailable)],
		"cars_rented":      counts[string(models.CarRented)],
		"cars_reserved":    counts[string(models.CarReserved)],
		"cars_maintenance": counts[string(models.CarMaintenance)],
		"today_revenue":    pricing.Round2(todayRevenue),
		"monthly_revenue":  pricing.Round2(monthlyRevenue),
		"active_rentals":   activeRentals,
		"upcoming_returns": upcomingReturns,
	}
}

// RevenueByMonth returns revenue totals for the last `months` calendar months.
func (s *Service) RevenueByMonth(months int) []map[string]interface{} {
	if months <= 0 {
		months = 12
	}
	now := startOfMonth(time.Now())
	results := make([]map[string]interface{}, 0, months)
	for i := months - 1; i >= 0; i-- {
		monthStart := now.AddDate(0, -i, 0)
		nextMonth := monthStart.AddDate(0, 1, 0)
		amount := s.revenueBetween(monthStart, nextMonth)
		results = append(results, map[string]interface{}{
			"month":   monthStart.Format("2006-01"),
			"revenue": pricing.Round2(amount),
		})
	}
	return results
}

// TopRentedCars returns the most frequently rented cars.
func (s *Service) TopRentedCars(limit int) []map[string]interface{} {
	if limit <= 0 {
		limit = 5
	}
	type row struct {
		ID      uint
		Brand   string
		Model   string
		Plate   string
		Rentals int64
	}
	var rows []row
	s.DB.Model(&models.Car{}).
		Select("cars.id, cars.brand, cars.model, cars.plate_number as plate, count(rentals.id) as rentals").
		Joins("JOIN rentals ON rentals.car_id = cars.id").
		Group("cars.id").
		Order("rentals DESC").
		Limit(limit).
		Scan(&rows)

	out := make([]map[string]interface{}, 0, len(rows))
	for _, r := range rows {
		out = append(out, map[string]interface{}{
			"car":     fmt.Sprintf("%s %s (%s)", r.Brand, r.Model, r.Plate),
			"rentals": r.Rentals,
		})
	}
	return out
}

// RentalStatistics returns rental counts grouped by status.
func (s *Service) RentalStatistics() map[string]interface{} {
	stats := map[string]interface{}{}
	for _, st := range models.AllRentalStatuses() {
		stats[string(st)] = int64(0)
	}
	type row struct {
		Status string
		Count  int64
	}
	var rows []row
	s.DB.Model(&models.Rental{}).Select("status, count(id) as count").Group("status").Scan(&rows)
	for _, r := range rows {
		stats[r.Status] = r.Count
	}
	return stats
}
